src/engine/Server.cpp

```diff
@@ -665,20 +710,148 @@ Awaitable<void> Server::sendStreamableResponse(
 }
 
 // ____________________________________________________________________________
-boost::asio::awaitable<void> Server::processQuery(
+class NoSupportedMediatypeError : public std::runtime_error {
+ public:
+  explicit NoSupportedMediatypeError(std::string_view msg)
+      : std::runtime_error{std::string{msg}} {}
+};
+
+// ____________________________________________________________________________
+MediaType Server::determineMediaType(
+    const ad_utility::url_parser::ParamValueMap& params,
+    const ad_utility::httpUtils::HttpRequest auto& request) {
+  // The following code block determines the media type to be used for the
+  // result. The media type is either determined by the "Accept:" header of
+  // the request or by the URL parameter "action=..." (for TSV and CSV export,
+  // for QLever-historical reasons).
+  std::optional<MediaType> mediaType = std::nullopt;
+
+  // The explicit `action=..._export` parameter have precedence over the
+  // `Accept:...` header field
+  if (checkParameter(params, "action", "csv_export")) {
+    mediaType = MediaType::csv;
+  } else if (checkParameter(params, "action", "tsv_export")) {
+    mediaType = MediaType::tsv;
+  } else if (checkParameter(params, "action", "qlever_json_export")) {
+    mediaType = MediaType::qleverJson;
+  } else if (checkParameter(params, "action", "sparql_json_export")) {
+    mediaType = MediaType::sparqlJson;
+  } else if (checkParameter(params, "action", "turtle_export")) {
+    mediaType = MediaType::turtle;
+  } else if (checkParameter(params, "action", "binary_export")) {
+    mediaType = MediaType::octetStream;
+  }
+
+  std::string_view acceptHeader = request.base()[http::field::accept];
+
+  if (!mediaType.has_value()) {
+    mediaType = ad_utility::getMediaTypeFromAcceptHeader(acceptHeader);
```

Comment:
Contrary to what the signature suggests and the docstring state, this function never returns `std::nullopt`. If no valid (as per what QLever knows) media type is found an exception is thrown. Fixing this and the corresponding signatures is another PR. For now I added a TODO.